File: algorithms/search/binary_search.py
# Binary Search

array = [num for num in range(100)]

# Time Complexity:      O(log n)
# Space Complexity :    O(1)
# Number of Steps:      O(log)
def binary_search(array, target):
    left = 0
    right = len(array) - 1

    while left <= right:
        mid = (left + right) // 2
        # or mid = left + (right - left) // 2

        if array[mid] == target:
            return mid
        elif array[mid] > target:
            right = mid - 1
        else:
            left = mid + 1

    return -1


print("Answer: ", binary_search(array, 23))

File: cpp_algorithms/search/binary_search.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;
#define pb push_back

// Binary Search

// Time Complexity:      O(log n)
// Space Complexity :    O(1)
// Number of Steps:      O(log n)
int binary_search(vector<int> arr, int target) {
    int left = 0;
    int right = arr.size() - 1;

    while (left <= right) {
        int mid = (left + right) / 2;
        // or mid = left + (right - left) / 2;

        if (arr[mid] == target) {
            return mid;
        } else if (arr[mid] > target) {
            right = mid - 1;
        } else {
            left = mid + 1;
        }
    }

    return -1;
}
